xchange-lgo/src/main/java/info/bitrich/xchangestream/lgo/LgoStreamingTradeService.java

```diff
@@ -0,0 +1,287 @@
+        Order copy = order instanceof LimitOrder ? LimitOrder.Builder.from(order).build() : MarketOrder.Builder.from(order).build();
+        // because actual released version of xchange-core has buggy Builder.from methods
+        copy.setFee(order.getFee());
+        copy.setCumulativeAmount(order.getCumulativeAmount());
+        // https://github.com/knowm/XChange/pull/3163
+        return copy;
+    }
+
+    @Override
+    public Observable<UserTrade> getUserTrades(CurrencyPair currencyPair, Object... args) {
+        return getBatchOrderEvents(currencyPair)
+                .filter(lgoOrderEvent -> lgoOrderEvent.getType().equals("match"))
+                .map(matchEvent -> LgoAdapter.adaptUserTrade(currencyPair, (LgoMatchOrderEvent) matchEvent));
+    }
+
+    /**
+     * Receive all events for the selected currency pairs. Merges batch order events and ack (AFR)
+     * events.
+     */
+    public Observable<LgoOrderEvent> getAllOrderEvents(Collection<CurrencyPair> currencyPairs) {
+        Observable<LgoOrderEvent> ackObservable = getReceivedOrderEvents();
+        Optional<Observable<LgoOrderEvent>> batchObservable = currencyPairs.stream()
+                .map(this::getBatchOrderEvents)
+                .reduce(Observable::mergeWith);
+        return batchObservable.isPresent() ? ackObservable.mergeWith(batchObservable.get()) : ackObservable;
+    }
+
+    /**
+     * Get ack for your placed orders. "received" events indicate the orderId associated to your
+     * order, if you set a reference on order placement you will have it in the event. "failed" events
+     * indicate that the order could not be read or was invalid and not added to a batch.
+     */
+    public Observable<LgoOrderEvent> getReceivedOrderEvents() {
+        if (afrSubscription == null) {
+            createAfrSubscription();
+        }
+        return afrSubscription.share();
+    }
+
+    private void createAfrSubscription() {
+        final ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+        afrSubscription = streamingService
+                .subscribeChannel("afr")
+                .map(s -> mapper.readValue(s.toString(), LgoAckUpdate.class))
+                .map(LgoAckUpdate::getData)
+                .flatMap(Observable::fromIterable)
+                .map(LgoAdapter::adaptReceivedEvent);
+    }
+
+    /**
+     * Get all events of your orders happened during batch execution, for a currency pair. "pending"
+     * events indicate that the order was added to a batch and received by the execution engine.
+     * "invalid" events indicate that the order was not suitable for execution. "match" events
+     * indicate that the order did match against another order. "open" events indicate that the order
+     * entered the order book. "done" events indicate that the order was filled, canceled or
+     * rejected.
+     */
+    public Observable<LgoOrderEvent> getBatchOrderEvents(CurrencyPair currencyPair) {
```

Comment:
Rename to `getRawBatchOrderEvents` in line with convention

Reply:
https://github.com/bitrich-info/xchange-stream/pull/385/commits/bf18a67a4e7b047677722216dbbf794927610689